src/test/java/org/jenkinsci/plugins/gitclient/JGitLightweightTagTest.java

```diff
@@ -0,0 +1,145 @@
+        repoRootGitDir = gitClient.withRepository((r, channel) -> r.getDirectory());
+        gitClient.init_()
+                 .workspace(repoRoot.getAbsolutePath())
+                 .execute();
+        assertThat(repoRootGitDir, is(anExistingDirectory()));
+    }
+
+    private ObjectId commitFile(final String path, final String content, final String commitMessage) throws Exception {
+        createFile(path, content);
+        gitClient.add(path);
+        gitClient.commit(commitMessage);
+
+        List<ObjectId> headList = gitClient.revList(HEAD);
+        assertThat(headList.size(), is(greaterThan(0)));
+        return headList.get(0);
+    }
+
+    private void createFile(final String path, final String content) throws Exception {
+        File aFile = new File(repoRoot, path);
+        File parentDir = aFile.getParentFile();
+        if (parentDir != null) {
+            parentDir.mkdirs();
+        }
+        try (PrintWriter writer = new PrintWriter(aFile, "UTF-8")) {
+            writer.printf(content);
+        }
+    }
+
+    private void packRefs() throws IOException {
+        try (FileRepository repo = new FileRepository(repoRootGitDir)) {
+            org.eclipse.jgit.internal.storage.file.GC gc;
+            gc = new org.eclipse.jgit.internal.storage.file.GC(repo);
+            gc.packRefs();
+        }
+    }
+
+    // No flavor of GitClient has a tag(String) API, only tag(String,String).
+    // But sometimes we want a lightweight a.k.a. non-annotated tag.
+    private void lightweightTag(String tagName) throws Exception {
+        try (FileRepository repo = new FileRepository(repoRootGitDir)) {
+            // Collides with implicit org.jenkinsci.plugins.gitclient.Git.
+            org.eclipse.jgit.api.Git jgitAPI = org.eclipse.jgit.api.Git.wrap(repo);
+            jgitAPI.tag().setName(tagName).setAnnotated(false).call();
+        }
+    }
+
+    @Issue("JENKINS-57205") // NPE on PreBuildMerge with packed lightweight tag
+    @Test
+    public void testGetTags_packedRefs() throws Exception {
+        // JENKINS-57205 is triggered by lightweight tags
+        ObjectId firstCommit = commitFile(
+            "first.txt",
+            "Great info here",
+            "First commit"
+        );
+        String lightweightTagName = "lightweight_tag";
+        lightweightTag(lightweightTagName);
+
+        // But throw in an annotated tag for symmetry and coverage
+        ObjectId secondCommit = commitFile(
+            "second.txt",
+            "Great info here, too",
+            "Second commit"
+        );
+        String annotatedTagName = "annotated_tag";
+        gitClient.tag(annotatedTagName, "Tag annotation");
+
+        packRefs();
+
+        Set<GitObject> tags = gitClient.getTags();
+
+        assertThat(
+            tags,
+            hasItem(
+                allOf(
+                    hasProperty("name", equalTo(lightweightTagName)),
+                    hasProperty("SHA1", equalTo(firstCommit))
```

Comment:
@brianeray can you help me understand how this assertion operates?  The field name inside the GitObject is `sha1` but the assertion is checking for a `SHA1` property.  When I tried changing it to `sha1`, the tests failed to find the property.  Clearly I have more to learn about writing good Hamcrest assertions.

Reply:
That lowercase `sha1` is exactly how I initially coded it, and it failed the same way you experienced.

Then I scrolled down that field's accessor, which [has different capitalization past the first letter](https://github.com/jenkinsci/git-client-plugin/blob/0ef4d20d057714bc5da618793b1c0c1c7bc9258e/src/main/java/hudson/plugins/git/GitObject.java#L38-L40). IIRC this particular Hamcrest matcher operates against the Java bean API, and thus goes through the getter instead straight at the field.

So it's sort of an accident of mismatched letter case in the bean here rather than idiosyncracy of Hamcrest. I think.